File: Julia/extended zip.jl
function myAllPairs(xs,ys)
    if xs == [] || ys == [] return [] end
    trim([pairs(popfirst!(xs), copy(ys)); myAllPairs(xs, ys)])
end

function pairs(x,ys)
    if ys == [] return [] end
    [(x, popfirst!(ys)); pairs(x,ys)]
end

function trim(xs)
    if xs == [] return [] end
    x = popfirst!(xs)
    append!([x],[i for i in trim(xs) if i != x])
end

function main()
    println(myAllPairs([],[]))
    println(myAllPairs([],[4,5,6]))
    println(myAllPairs([1,2,3],[]))
    println(myAllPairs([1, 2, 3], [4, 5, 6]))
    println(myAllPairs([1, 3, 3], [3, 5, 6]))
end

main()

File: Prolog/extended zip.pl
myAllPairs([],_,[]) :- !.
myAllPairs(_,[],[]) :- !.
myAllPairs([H|T],Y,L) :-
        pairs(H,Y,L1),!,
        myAllPairs(T,Y,L2),!,
        merge(L1,L2,L3),!,
        trim(L3,L).

pairs(_,[],[]) :- !.
pairs(X,[Y|T],[(X,Y)|L]) :-
        pairs(X,T,L).

merge([],L,L) :- !.
merge([H|T],L,[H|L1]) :-
        merge(T,L,L1).

trim([],[]) :- !.
trim([H|T],L) :-
        trim(T,L1),
        (   \+ exist(H,L1)
        ->  L = [H|L1]
        ;   L = L1
        ).

exist(E,[E|_]) :- !.
exist(E,[_|L]) :-
        exist(E,L).